src/main/kotlin/org/rust/ide/refactoring/inlineTypeAlias/RsInlineTypeAliasProcessor.kt

```diff
@@ -0,0 +1,172 @@
+            RsImportHelper.importTypeReferencesFromElement(context, typeReference)
+        }
+    }
+
+    override fun getCommandName(): String = "Inline Type Alias ${typeAlias.name}"
+
+    override fun createUsageViewDescriptor(usages: Array<UsageInfo>): UsageViewDescriptor =
+        RsInlineUsageViewDescriptor(typeAlias, "Type Alias to inline")
+}
+
+/**
+ * type Foo<T> = Vec<T>;
+ * fn main() {
+ *     let v = Foo::new();
+ *             ~~~ has substitution `T => T`
+ *             ~~~~~~~~ has substitution `Self => Vec<i32>`
+ *     v.push(1);
+ * }
+ */
+fun tryGetTypeAliasSubstitutionUsingParent(path: RsPath, typeAlias: RsTypeAlias): Substitution? {
+    val parentPath = path.parent as? RsPath ?: return null
+    if (parentPath.parent !is RsPathExpr) return null
+    val resolvedMethod = parentPath.reference?.advancedResolve() ?: return null
+    val selfTy = resolvedMethod.subst[TyTypeParameter.self()] ?: return null
+    val inference = path.implLookup.ctx
+    val subst = inference.instantiateBounds(typeAlias, selfTy)
+    val typeReference = typeAlias.typeReference ?: return null
+    val type = typeReference.normType.substitute(subst)
+    inference.combineTypes(type, selfTy)
+    return subst.mapTypeValues { (_, v) -> inference.resolveTypeVarsIfPossible(v) }
+}
+
+/**
+ * type Foo<T> = Vec<T>;
+ *               ~~~~~~ type
+ * fn main() {
+ *     let _: Foo<i32>;
+ *            ~~~~~~~~ path
+ * }                   subst = "T => i32"
+ */
+fun replacePathWithType(path: RsPath, typeReference0: RsTypeReference, substitution: Substitution): RsElement? {
+    val factory = RsPsiFactory(path.project)
+    val typeText = typeReference0.getStubOnlyText(substitution, shortPaths = false)
+    val typeReference = factory.tryCreateType(typeText) ?: return null
+    val typeAsPath = (typeReference as? RsPathType)?.path
+
+    val parent = path.parent
+    return when {
+        parent is RsTypeReference -> {
+            parent.replace(typeReference)
+        }
+
+        typeAsPath != null -> {
+            if (path.ancestorStrict<RsTypeReference>() == null) {
+                // Consider `type Foo = HashSet<i32>;`
+                // Replace `Foo::new()` to `HashSet::<i32>::new()`
+                typeAsPath.typeArgumentList?.addAfter(factory.createColonColon(), null)
+            }
+
+            path.replace(typeAsPath)
+        }
+
+        parent is RsPath -> {
+            // Consider `type Foo = [i32];`
+            // Replace `Foo::is_empty(x)` to `<[i32]>::is_empty(x)`
+            val parentName = parent.referenceName ?: return null
+            val parentTypeArguments = parent.typeArgumentList?.text.orEmpty()
+            val parentNewText = "<$typeText>::$parentName$parentTypeArguments"
+            val parentNew = factory.tryCreatePath(parentNewText) ?: return null
+            parent.replace(parentNew)
+        }
+
+        // should be no other cases
+        else -> return null
```

Comment:
Since we strictly cast it to `RsElement` later potentially throwing NPE, I think it's better to fail immediately with some descriptive error message

Reply:
This branch will not result in `NPE`, we return `null` which will be later ignored by `mapNotNull`

Reply:
Oh, you are right, thanks!